python/cuspatial/cuspatial/core/binpreds/feature_contains.py

```diff
@@ -87,47 +99,35 @@ def _compute_polygon_polygon_contains(self, lhs, rhs, preprocessor_result):
-        # Divide the linestring in half and test the point for containment
-        # in the polygon.
-        size_two = rhs.sizes == 2
-        if (size_two).any():
-            center_points = _linestrings_to_center_point(rhs[size_two])
-            size_two_results = _false_series(len(lhs))
-            size_two_results.iloc[rhs.index[size_two]] = (
-                _basic_contains_count(lhs, center_points) > 0
-            )
-            return size_two_results
-        else:
-            return _false_series(len(lhs))
-
     def _compute_polygon_linestring_contains(
         self, lhs, rhs, preprocessor_result
     ):
+        # Count the number of points in lhs that are properly contained by
+        # rhs
         contains = _basic_contains_count(lhs, rhs).reset_index(drop=True)
-        intersects = self._intersection_results_for_contains(lhs, rhs)
-
-        # If a linestring has intersection but not containment, we need to
-        # test if the linestring is in the interior of the polygon.
-        final_result = _false_series(len(lhs))
-        intersection_with_no_containment = (contains == 0) & (intersects != 0)
-        interior_tests = self._test_interior(
-            lhs[intersection_with_no_containment].reset_index(drop=True),
-            rhs[intersection_with_no_containment].reset_index(drop=True),
+
+        # Count the number of point intersections (line crossings) between
+        # lhs and rhs.
+        # Also count the number of perfectly overlapping linestring sections.
+        # Each linestring overlap counts as two point overlaps.
+        (
+            point_intersects_count,
+            linestring_intersects_count,
+        ) = self._intersection_results_for_contains_linestring(lhs, rhs)
+
+        # Subtract the length of the linestring intersections from the length
+        # of the rhs linestring, then test that the sum of contained points
+        # is equal to that adjusted rhs length.
+        rhs_sizes_less_line_intersection_size = (
+            rhs.sizes - linestring_intersects_count
         )
-        interior_tests.index = intersection_with_no_containment[
-            intersection_with_no_containment
-        ].index
-        # LineStrings that have intersection but no containment are set
-        # according to the `intersection_with_no_containment` mask.
-        final_result[intersection_with_no_containment] = interior_tests
-        # LineStrings that do not are contained if the sum of intersecting
-        # and containing points is greater than or equal to the number of
-        # points that make up the linestring.
-        final_result[~intersection_with_no_containment] = (
-            contains + intersects >= rhs.sizes
+        rhs_sizes_less_line_intersection_size[
+            rhs_sizes_less_line_intersection_size <= 0
+        ] = 1
+        final_result = contains + point_intersects_count == (
+            rhs_sizes_less_line_intersection_size
```

Comment:
This seems differ from your statement in the PR description, your description basically says:
```python
final_result = contains == rhs.sizes - linestring_points_on_boundary
```

To make this equal to the formula here, it seems like we need:
```
linestring_points_on_boundary == linestring_intersects_counts - point_intersects_count
```

Though I don't see the connection here. Can you show me some hints?

Reply:
In the description I believe I misspoke about the number of points that are contained in the polygon as demonstrated in the code:

```
final_result = contains + point_intersects_count == rhs.sizes - linestring_points_on_boundary
```

IIRC, you are concerned that `linestring_points_on_boundary` in your example doesn't take into account the number of point intersections, which are ignored in the description of this PR but not in `_compute_polygon_linestring_contains`.

`contains` equals the number of the LineString vertices that are contained by the polygon.

`point_intersects_count` equals the number of LineString vertices that intersect with the polygon: that is, the places where the LineString does not share an overlapping boundary with the polygon, but instead has a vertex that is colinear with the boundary of the polygon, but each segment that shares the vertex is not colinear. When a `point_intersects_count` intersection occurs, there must either: exist a point that is inside the polygon that will be counted by `contains` or a point outside of the polygon that will not be counted. 

When a vertex is colinear with the boundary of the polygon this is counted as "in" the polygon for a reason related to our above conversation, where the minimum LineString length must be 1. Consider the following example:

![image](https://github.com/rapidsai/cuspatial/assets/410931/3b70b5af-8fe0-4754-89a5-6d4075cd2d86)

In the first case, `point_intersects_count == 2`, as the LineString vertices are colinear with the boundary of the polygon. `contains + point_intersects_count == rhs.sizes` and the LineString is contained.

In the second case, `point_intersects_count == 0`, as the LineString vertices are not colinear- it is a genuine intersection. `contains + point_intersects_count < rhs.sizes` and the LineString is not contained.

`linestring_intersects_count` is the number of LineString vertices that are from segments that are colinear with the polygon boundary. They are used only to reduce the number of points in the LineString that should be tested for .contains.

The final equation would be
```final_result = contains + point_intersects_count == rhs.sizes - linestring_intersects_count```

equal to

```final_result = contains == rhs.sizes - linestring_intersects_count - point_intersects_count``` as I think you are alluding to above, except that the LineString must maintain a size of at least 0 as you asked about above. Maybe I'm missing something here that would allow me to subtract `point_intersects_count` instead of using the line that sets the minimum `rhs_sizes_less_line_intersection_size` to 1. I'll be thinking of that.

Consider the final case in this response, 

![image](https://github.com/rapidsai/cuspatial/assets/410931/966cf8f8-c0c6-4853-84fb-58b4d865398d)

In this case `point_intersects_count = 1`, and due to the overlapping segment, `rhs_sizes_less_line_intersection_count == 1`, and the LineString is contained. I could do `contains == rhs.sizes - linestring_intersects_count - point_intersects_count`, which in this case would be `(0 == 0)`, but that would break the third case in my earlier response, where the LineString is entirely colinear with the boundary of the polygon. In those cases the LineString is not contained, and as such it must be that `rhs_sizes_less_line_intersection_count > 0`

Reply:
I see you arrive at the same equality as I did above. 
```
final_result = contains == rhs.sizes - linestring_intersects_count - point_intersects_count
```

Great. I'm still confused with the discrepency between this and code. But perhaps focus on Mark's question below first.